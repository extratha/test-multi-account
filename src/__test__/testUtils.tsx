import { RenderOptions, RenderResult, render } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import { SWRConfig } from 'swr';
import { theme } from '@/config/config-mui/theme';
import { ThemeProvider as MUIThemeProvider } from '@mui/material/styles';
import { NextIntlClientProvider } from 'next-intl';
import commonTh from '../../public/locales/th/common.json';
import commonEn from '../../public/locales/th/common.json';
type RenderWithProvider = (
  elm: React.ReactElement,
  renderOptions?: ProviderOptions,
) => RenderResult;

export type ProviderOptions = {
  formOptions?: Record<string, unknown>;
} & RenderOptions;

const messages = {
  ...commonTh,
  ...commonEn,
};

const renderWithProviders: RenderWithProvider = (
  component,
  renderOptions?,
  locale = 'th',
) => {
  const wrapper = ({ children }: { children: React.ReactNode }) => (
    <NextIntlClientProvider locale={locale} messages={messages}>
      <MUIThemeProvider theme={theme}>
        <SWRConfig value={{ provider: () => new Map() }}>{children}</SWRConfig>
      </MUIThemeProvider>
    </NextIntlClientProvider>
  );

  return render(component, {
    wrapper,
    ...renderOptions,
  });
};

export { renderWithProviders as render, userEvent };
export * from '@testing-library/react';
