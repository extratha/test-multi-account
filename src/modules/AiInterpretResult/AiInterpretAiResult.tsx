import { Button, Stack, Typography } from "@mui/material";
import { styled } from "@mui/material/styles";
import { useMemo, useState } from "react";
import Markdown from "react-markdown";

import { IconCopy, IconGlobe } from "@/assets";
import { LANGUAGE } from "@/constant";
import useTranslation from "@/locales/useLocale";
import { AiData } from "@/types/model.api";

export interface AiInterpretAiResultProps {
  name: string;
  title: string;
  data: AiData;
}

const TitleResult = styled(Typography)(({ theme }) => ({
  fontWeight: 600,
  color: theme.palette.primary.main,
}));

const AiInterpretAiResult = ({ name, title, data }: AiInterpretAiResultProps) => {
  const { translation } = useTranslation();
  const [language, setLanguage] = useState(LANGUAGE.TH);

  const description = useMemo(() => {
    const text = data.descriptions.find((item) => item.language === language);
    return text?.description || "";
  }, [language]);

  const handleClickTranslate = () => {
    setLanguage(language === LANGUAGE.TH ? LANGUAGE.EN : LANGUAGE.TH);
  };

  const handleClickCopy = (text: string) => {
    navigator.clipboard.writeText(`${data.title}\n${text}`);
  };

  return (
    <Stack spacing="16px">
      <TitleResult data-testid={`${name}-title`}>{title}</TitleResult>
      <Markdown data-testid={`${name}-description`}>{description}</Markdown>
      <Stack direction="row" spacing="16px">
        <Button
          startIcon={<IconCopy />}
          onClick={() => handleClickCopy(description)}
          data-testid={`${name}-button-copy`}
        >
          {translation("AiInterpret.aiInterpretResult.button.copy")}
        </Button>
        {data.descriptions.length > 1 && (
          <Button
            startIcon={<IconGlobe />}
            onClick={handleClickTranslate}
            data-testid={`${name}-change-translate-button`}
          >
            {translation("AiInterpret.aiInterpretResult.button.multiple")}
          </Button>
        )}
      </Stack>
    </Stack>
  );
};

export default AiInterpretAiResult;
