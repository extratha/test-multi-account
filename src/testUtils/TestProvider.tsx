import { ThemeProvider } from "@emotion/react";

import { ReactNode } from "react";
import { SWRConfig } from "swr";

import ModalLayer from "@/components/Modal/ModalLayer";
import { theme } from "@/config/config-mui";

interface TestProviderProps {
  children?: ReactNode;
}

export const TestProvider = ({ children }: TestProviderProps) => {
  return (
    <ThemeProvider theme={theme}>
      <SWRConfig value={{ provider: () => new Map() }}>{children}</SWRConfig>
      <ModalLayer />
    </ThemeProvider>
  );
};
