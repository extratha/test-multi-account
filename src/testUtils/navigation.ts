import * as NextNavigation from "next/navigation";

jest.mock("next/navigation", () => ({
  useRouter: jest.fn(),
  useParams: jest.fn(),
  usePathname: jest.fn(),
  useSearchParams: jest.fn(),
  useServerInsertedHTML: jest.fn(),
}));

export interface SpyUseSearchParams {
  get: jest.Mock;
}

export interface SpyUseRouter {
  replace: jest.Mock;
  push: jest.Mock;
  back: jest.Mock;
}

export const spyUseParams = (params: any = {}) => {
  jest.spyOn(NextNavigation, "useParams").mockReturnValue(params);
  return NextNavigation.useParams as jest.Mock;
};

export const spyUseRouter = (): SpyUseRouter => {
  const replace = jest.fn();
  const push = jest.fn();
  const back = jest.fn();
  jest.spyOn(NextNavigation, "useRouter").mockReturnValue({ replace, push, back } as any);
  return { replace, push, back };
};

export const spyUsePathname = (pathname: string) => {
  jest.spyOn(NextNavigation, "usePathname").mockReturnValue(pathname);
};

export const spyUseSearchParams = (): SpyUseSearchParams => {
  const get = jest.fn();
  jest.spyOn(NextNavigation, "useSearchParams").mockReturnValue({ get } as any);
  return {
    get,
  };
};

export const spyUseServerInsertedHTML = () => {
  return jest.spyOn(NextNavigation, "useServerInsertedHTML");
};
