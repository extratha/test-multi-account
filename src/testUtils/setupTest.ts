import "@testing-library/jest-dom";
import "jest-canvas-mock";

import "@/__mocks__/Markdown";
import "@/__mocks__/MockFirebase";
import "@/__mocks__/NextImage";
import { LOCALE_LANGUAGE } from "@/config";
import { initI18next } from "@/locales/i18n";
import locale from "@/locales/th.json";

initI18next(LOCALE_LANGUAGE.TH, locale);

jest.mock("@/api/api", () => {
  const actualModule = jest.requireActual("@/api/api");
  return { ...actualModule, __esModule: true };
});

jest.mock("@/api/apiUnauthorize", () => {
  const actualModule = jest.requireActual("@/api/apiUnauthorize");
  return { ...actualModule, __esModule: true };
});

jest.mock("next/config", () => () => ({
  setConfig: jest.fn(),
  publicRuntimeConfig: {
    apiToken:
      "eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJleHAiOjE3MDY2NzUzNDQsImlhdCI6MTcwNjU4ODk0NCwicHJvZmlsZUlkIjoiZDU1YWY3NDUtOTYyZS00MTk2LTkyOTEtZDg2YTliYWNiZWY2In0.sNgwYjIGLYEVAptqILGgRVGXwt9FCPKHRubEZyaodhg",
    mockBaseApiUrl: "https://www.mock-playground-dashboard.co.th",
    baseApiUrl: "https://www.mock-playground-dashboard.co.th",
    globalEndpoint: "playground-dashboard",
    publicApiHost: "http://localhost:3000/",
  },
}));

jest.mock("cookies-next", () => ({
  __esModule: true,
  ...jest.requireActual("cookies-next"),
  getCookie: jest.fn().mockReturnValue("MOCK_ACCESS_TOKEN"),
}));
