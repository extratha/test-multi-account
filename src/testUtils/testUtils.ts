import { act, render as testRender } from "@testing-library/react";
import dayjs from "dayjs";
import flushPromises from "flush-promises";
import { ReactElement } from "react";

import { TestProvider } from "./TestProvider";

export const API = {
  LOGIN: "/auth/login",
  EXAMPLE_AI_INTERPRET_URL: "/lab/examples/labId",
  GET_LAB_INTERPRETS_ID_URL: "/lab/interprets/labId",
  SUBMIT_LAB_INTERPRETS_URL: "/lab/interprets",
  CONSENT: "/consents",
  TERMS_AND_CONDITIONS: "/consents/terms-conditions",
  TERMS_AND_CONDITIONS_LATEST: "/consents/terms-conditions/latest",
  PRIVACY_POLICY: "/consents/privacy-policies",
  PRIVACY_POLICY_LATEST: "/consents/privacy-policies/latest",
  PUBLIC_TERMS_CONDITIONS: "/consents/terms-conditions/public",
  PUBLIC_PRIVACY_POLICIES: "/consents/privacy-policies/public",
};

export const render = (ui: ReactElement) => {
  return testRender(ui, { wrapper: TestProvider });
};

export const flushPromise = async () => {
  await act(async () => {
    await flushPromises();
  });
};

export const advanceTimersByTime = async (time: number) => {
  await act(async () => {
    jest.advanceTimersByTime(time);
  });
};

export const setMockDate = (date: string) => {
  jest.setSystemTime(dayjs(date).valueOf());
};
