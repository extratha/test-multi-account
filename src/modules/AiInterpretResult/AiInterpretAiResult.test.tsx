import { render } from "@/testUtils/testUtils";
import { screen } from "@testing-library/react";
import userEvent from "@testing-library/user-event";

import AiInterpretAiResult, { AiInterpretAiResultProps } from "./AiInterpretAiResult";

describe("AiInterpretAiResult", () => {
  let props: AiInterpretAiResultProps;

  beforeEach(() => {
    props = {
      name: "ai-interpret-0",
      title: "title general",
      data: {
        title: "title general",
        descriptions: [
          {
            description: "language th",
            language: "TH",
          },
          {
            description: "language en",
            language: "EN",
          },
        ],
      },
    };

    Object.assign(navigator, {
      clipboard: {
        writeText: jest.fn(),
      },
    });
  });

  const renderAiInterpretAiResult = () => {
    return render(<AiInterpretAiResult {...props} />);
  };

  it("should render correctly", () => {
    const { asFragment } = renderAiInterpretAiResult();
    expect(asFragment()).toMatchSnapshot();
  });

  it("should switch language when click switch language button", async () => {
    renderAiInterpretAiResult();

    expect(screen.getByTestId("ai-interpret-0-description")).toHaveTextContent("language th");

    await userEvent.click(screen.getByTestId("ai-interpret-0-change-translate-button"));

    expect(screen.getByTestId("ai-interpret-0-description")).toHaveTextContent("language en");

    await userEvent.click(screen.getByTestId("ai-interpret-0-change-translate-button"));

    expect(screen.getByTestId("ai-interpret-0-description")).toHaveTextContent("language th");
  });

  it("should clipboard ai data when click copy button", async () => {
    renderAiInterpretAiResult();

    await userEvent.click(screen.getByTestId("ai-interpret-0-button-copy"));

    expect(navigator.clipboard.writeText).toHaveBeenCalledWith("title general\nlanguage th");
  });

  it("should hide switch language button when ai data has only language en", async () => {
    props.data.descriptions = [
      {
        description: "language en",
        language: "EN",
      },
    ];

    renderAiInterpretAiResult();

    expect(screen.queryByTestId("ai-interpret-0-change-translate-button")).not.toBeInTheDocument();
  });
});
