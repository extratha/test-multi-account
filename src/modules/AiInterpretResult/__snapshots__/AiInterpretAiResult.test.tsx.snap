// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AiInterpretAiResult should render correctly 1`] = `
<DocumentFragment>
  .emotion-0 {
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
}

.emotion-0>:not(style):not(style) {
  margin: 0;
}

.emotion-0>:not(style)~:not(style) {
  margin-top: 16px;
}

.emotion-1 {
  margin: 0;
  font-family: fontFamily;
  font-weight: 600;
  font-size: 1rem;
  line-height: 1.5;
  color: #2196F3;
}

.emotion-2 {
  display: flex;
  -webkit-flex-direction: row;
  -ms-flex-direction: row;
  flex-direction: row;
}

.emotion-2>:not(style):not(style) {
  margin: 0;
}

.emotion-2>:not(style)~:not(style) {
  margin-left: 16px;
}

.emotion-3 {
  display: inline-flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -ms-flex-pack: center;
  -webkit-justify-content: center;
  justify-content: center;
  position: relative;
  box-sizing: border-box;
  -webkit-tap-highlight-color: transparent;
  background-color: transparent;
  outline: 0;
  border: 0;
  margin: 0;
  border-radius: 4px;
  padding: 6px 8px;
  cursor: pointer;
  -webkit-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  user-select: none;
  vertical-align: middle;
  -moz-appearance: none;
  -webkit-appearance: none;
  -webkit-text-decoration: none;
  text-decoration: none;
  color: #2196F3;
  font-family: fontFamily;
  font-weight: 700;
  font-size: 12px;
  line-height: 1.5;
  text-transform: uppercase;
  min-width: 64px;
  -webkit-transition: background-color 250ms cubic-bezier(0.4, 0, 0.2, 1) 0ms,box-shadow 250ms cubic-bezier(0.4, 0, 0.2, 1) 0ms,border-color 250ms cubic-bezier(0.4, 0, 0.2, 1) 0ms,color 250ms cubic-bezier(0.4, 0, 0.2, 1) 0ms;
  transition: background-color 250ms cubic-bezier(0.4, 0, 0.2, 1) 0ms,box-shadow 250ms cubic-bezier(0.4, 0, 0.2, 1) 0ms,border-color 250ms cubic-bezier(0.4, 0, 0.2, 1) 0ms,color 250ms cubic-bezier(0.4, 0, 0.2, 1) 0ms;
}

.emotion-3::-moz-focus-inner {
  border-style: none;
}

.emotion-3.Mui-disabled {
  pointer-events: none;
  cursor: default;
}

@media print {
  .emotion-3 {
    -webkit-print-color-adjust: exact;
    color-adjust: exact;
  }
}

.emotion-3:hover {
  -webkit-text-decoration: none;
  text-decoration: none;
  background-color: rgba(33, 150, 243, 0.04);
}

@media (hover: none) {
  .emotion-3:hover {
    background-color: transparent;
  }
}

.emotion-3.Mui-disabled {
  color: rgba(0, 0, 0, 0.26);
}

.emotion-4 {
  display: inherit;
  margin-right: 8px;
  margin-left: -4px;
}

.emotion-4>*:nth-of-type(1) {
  font-size: 20px;
}

.emotion-5 {
  overflow: hidden;
  pointer-events: none;
  position: absolute;
  z-index: 0;
  top: 0;
  right: 0;
  bottom: 0;
  left: 0;
  border-radius: inherit;
}

<div
    class="MuiStack-root emotion-0"
  >
    <p
      class="MuiTypography-root MuiTypography-body1 emotion-1"
      data-testid="ai-interpret-0-title"
    >
      title general
    </p>
    <div
      data-testid="ai-interpret-0-description"
    >
      language th
    </div>
    <div
      class="MuiStack-root emotion-2"
    >
      <button
        class="MuiButtonBase-root MuiButton-root MuiButton-text MuiButton-textPrimary MuiButton-sizeMedium MuiButton-textSizeMedium MuiButton-colorPrimary MuiButton-root MuiButton-text MuiButton-textPrimary MuiButton-sizeMedium MuiButton-textSizeMedium MuiButton-colorPrimary emotion-3"
        data-testid="ai-interpret-0-button-copy"
        tabindex="0"
        type="button"
      >
        <span
          class="MuiButton-icon MuiButton-startIcon MuiButton-iconSizeMedium emotion-4"
        >
          <svg />
        </span>
        คัดลอกข้อความ
        <span
          class="MuiTouchRipple-root emotion-5"
        />
      </button>
      <button
        class="MuiButtonBase-root MuiButton-root MuiButton-text MuiButton-textPrimary MuiButton-sizeMedium MuiButton-textSizeMedium MuiButton-colorPrimary MuiButton-root MuiButton-text MuiButton-textPrimary MuiButton-sizeMedium MuiButton-textSizeMedium MuiButton-colorPrimary emotion-3"
        data-testid="ai-interpret-0-change-translate-button"
        tabindex="0"
        type="button"
      >
        <span
          class="MuiButton-icon MuiButton-startIcon MuiButton-iconSizeMedium emotion-4"
        >
          <svg />
        </span>
        See Translate
        <span
          class="MuiTouchRipple-root emotion-5"
        />
      </button>
    </div>
  </div>
</DocumentFragment>
`;
